### PR TITLE
Revert "Revert "Revert "Revert "Balder/no more fs4 dispatching from fastsearcher"""".

### PR DESCRIPTION
**Pull request by baldersheim**
Reverts vespa-engine/vespa#10693
To be continued

### CODE REVIEW

**Comment by baldersheim** [comment]
@bratseth PR
It should now be ready for another run through system tests.

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchCluster.java [hunk 1]
```diff
@@ -303,7 +303,7 @@ public void pingIterationCompleted() {
             updateSufficientCoverage(group, sufficientCoverage);
             trackGroupCoverageChanges(i, group, sufficientCoverage, averageDocumentsInOtherGroups);
         }
-        if ( ! anyGroupsSufficientCoverage ) {
+        if ( ! anyGroupsSufficientCoverage && (sumOfActiveDocuments == 0)) {
             vipStatus.removeFromRotation(clusterId);
```
Comment (bratseth) [inline container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchCluster.java]: I casn't follow this, why did you change this logic? It used to be quite clean and symmetric and now I'm confused.

**Review by bratseth — COMMENTED** [review]
I remember now that the distinction between prelude.ClusterController and search.ClusterController is that the former is used for all internal content clusters while the latter is for separate services. It would be nice to unify this and get rid of the former but I suggest we don't do it as part of this large diff. The logic in the prelude.ClusterController etc. is the end result of many iterations of problem solving and I worry that we'll create problems for scenarios not well covered by tests. How did we end up removing that as part of this? 

I guess it's fine to get another through factory but I'm quite confused by this and we should talk about it tomorrow when my brain works again :-)

**Comment by baldersheim** [comment]
I assume that when you say prelude.ClusterController, you mean prelude.ClusterMonitor.
prelude.ClusterMonitor went away as it used ping over fs4.
search.ClusterMonitor is used by both search.ClusterSearcher (for external http based clusters), and search.dispatch.SearchCluster which is the rpc based dispatch.
Currently both prelude.ClusterMonitor and dispatch.ClusterMonitor has been monitoring the backends. When I removed monitoring of fs4 the hierarchic_distribution test started failing.
dispatch.SearchCluster which is the NodeManager for dispatch.ClusterMonitor would only control the vipstatus if it was a direct dispatch group/node. And would return 404 when 200 was expected.
After I got it to atleast put it into rotation, then I also had to make get it to take it out of rotation when all nodes in all groups where down.

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchCluster.java (hunk 1 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchCluster.java]: I have added some clarifying comments to the code. But in essence this will remove it from rotation when no groups have sufficient coverage and there are no documents to be served. In essence this is when all nodes in all groups are down.

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchCluster.java [hunk 2]
```diff
@@ -229,7 +229,7 @@ private void updateSufficientCoverage(Group group, boolean sufficientCoverage) {
         group.setHasSufficientCoverage(sufficientCoverage);
         if ((!isInRotation || isDirectDispatchGroupAndChange) && sufficientCoverage) {
             vipStatus.addToRotation(clusterId);
-        } else if (isDirectDispatchGroupAndChange && ! sufficientCoverage) {
+        } else if (isDirectDispatchGroupAndChange) {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchCluster.java]: Not sure if I follow. Is this with respect to readability, or correctness ?
The rewrite here was to enable it put it in rotation if it was not also when it is not a direct dispatch group. If not status would never be 'green'.
